src/hotspot/share/utilities/globalDefinitions.hpp

```diff
@@ -1105,7 +1106,22 @@ template<class T> constexpr T MIN3(T a, T b, T c)      { return MIN2(MIN2(a, b),
 template<class T> constexpr T MAX4(T a, T b, T c, T d) { return MAX2(MAX3(a, b, c), d); }
 template<class T> constexpr T MIN4(T a, T b, T c, T d) { return MIN2(MIN3(a, b, c), d); }
 
-template<class T> inline T ABS(T x)                 { return (x > 0) ? x : -x; }
+#define ABS(x) asserted_abs(x, __FILE__, __LINE__)
+
+template<class T> inline T asserted_abs(T x, const char* file, int line) {
+#ifdef ASSERT
+  if (std::is_integral<T>::value && x == std::numeric_limits<T>::min()) {
+    report_vm_error(file, line, "ABS: argument should not allow overflow");
+  }
+#endif
+  T res = (x > 0) ? x : -x;
+#ifdef ASSERT
+  if (res < 0) {
```

Comment:
I don't see how we could ever hit this.  Checking input seems sufficient.

Reply:
I thought we would like to check for other (unknown) cases for ABS failures, since we would otherwise presume only the `ABS(min_int)` is problematic. But I can remove it, and we would add more cases later if we find them. See new commit.